program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,371 @@ use crate::c_oracle_header::{
+                                      * weighted sums */
+    running_confidence:         [UnsignedTrackerRunningSum; NUM_ENTRIES], /* confidence running
+                                                                           * time
+                                                                           * weighted sums */
+    current_entry:              usize, //the current entry
+    max_observed_entry_abs_val: u64,   /* used to ensure that there are no overflows when
+                                        * looking at running sum diffs */
+    entry_validity:             [u8; NUM_ENTRIES], //entry validity
+}
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    SmaTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    ///invalidates current_entry, and increments self.current_entry num_entries times
+    fn invaldate_following_entries(&mut self, num_entries: usize) {
+        for _ in 0..num_entries {
+            self.entry_validity[self.current_entry] = 0;
+            self.current_entry = Self::get_next_entry(self.current_entry);
+        }
+    }
+    /// returns a weighted average of v with weights w
+    fn weighted_average<
+        T: std::ops::Mul<Output = T>
+            + std::ops::Add<Output = T>
+            + std::ops::Div<Output = T>
+            + std::marker::Copy,
+    >(
+        w1: T,
+        v1: T,
+        w2: T,
+        v2: T,
+    ) -> Result<T, OracleError> {
+        Ok((w1 * v1 + w2 * v2) / (w1 + w2))
+    }
+    ///Adds price and confidence each multiplied by update_time
+    /// to the current entry
+    fn add_price_to_current_entry(
+        &mut self,
+        prev_price: i64,
+        prev_conf: u64,
+        current_price: i64,
+        current_conf: u64,
+        update_time: i64,
+    ) -> Result<(), OracleError> {
+        let avg_price = (prev_price + current_price) / 2;
+        let avg_conf = (prev_conf + current_conf) / 2;
+        self.running_price[self.current_entry] +=
+            try_convert::<i64, SignedTrackerRunningSum>(update_time * avg_price)?;
+        self.running_confidence[self.current_entry] += try_convert::<u64, UnsignedTrackerRunningSum>(
+            try_convert::<i64, u64>(update_time)? * avg_conf,
+        )?;
+        //since price is always greater than confidence, it is sufficient to look at the
+        // price we also look at the average because it is the value that
+        // gets multiplied with the number of seconds
+        self.max_observed_entry_abs_val =
+            cmp::max(try_convert(avg_price)?, self.max_observed_entry_abs_val);
+        Ok(())
+    }
+}
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    Tracker<GRANUALITY, NUM_ENTRIES, THRESHOLD> for SmaTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    fn initialize(&mut self) -> Result<(), OracleError> {
+        //ensure that the first entry is invalid
+        self.max_update_time = THRESHOLD;
```

Comment:
Shouldn't the first entry always get invalidated because `max_update_time = current_time - 0`?

Reply:
No, that is not the case when you resize the account!